core ap_ssl_() support, related changes, latest mod_md

icing opened the pull request:
- adding new ap_ssl_*() functions for a backward
       compatible replacement of the major optional mod_ssl functions. This
       allows other ssl modules to work without impersonating mod_ssl and
       also allows different ssl modules being active on separate ports.
- latest mod_md with support for multiple certificates per domain
       and ECDSA certificates. Removed ACMEv1 support.
- Interworking mod_md and ssl modules changed to exchange PEM strings
    instead of file paths for ACME challenges.

icing committed:
Changed ap_ssl_answer_challenge() and its hook to provide PEM data for
certificate and key instead of file names.

Added support for this in mod_ssl and verified with a local mod_md
version that uses it.

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1887151 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
And the necessary log tags added just shortly afterwards.

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1887152 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
*) mod_md: v2.4.0 with improvements and bugfixes
     - MDPrivateKeys allows the specification of several types. Beside "RSA" plus
     optional key lengths elliptic curves can be configured. This means you can
     have multiple certificates for a Managed Domain with different key types.
     With ```MDPrivateKeys secp384r1 rsa2048``` you get one ECDSA  and one RSA
     certificate and all modern client will use the shorter ECDSA, while older
     client will get the RSA certificate.
     Many thanks to @tlhackque who pushed and helped on this.
     - Support added for MDomains consisting of a wildcard. Configuring
     ```MDomain *.host.net``` will match all virtual hosts matching that pattern
     and obtain one certificate for it (assuming you have 'dns-01' challenge
     support configured). Addresses #239.
     - Removed support for ACMEv1 servers. The only known installation used to
     be Let's Encrypt which has disabled that version more than a year ago for
     new accounts.
     - Andreas Ulm (<https://github.com/root360-AndreasUlm>) implemented the
     ```renewing``` call to ```MDMessageCmd``` that can deny a certificate
     renewal attempt. This is useful in clustered installations, as
     discussed in #233).
     - New event ```challenge-setup:<type>:<domain>```, triggered when the
     challenge data for a domain has been created. This is invoked before the
     ACME server is told to check for it. The type is one of the ACME challenge
     types. This is invoked for every DNS name in a MDomain.
     - The max delay for retries has been raised to daily (this is like all
     retries jittered somewhat to avoid repeats at fixed time of day).
     - Certain error codes reported by the ACME server that indicate a problem
     with the configured data now immediately switch to daily retries. For
     example: if the ACME server rejects a contact email or a domain name,
     frequent retries will most likely not solve the problem. But daily retries
     still make sense as there might be an error at the server and un-supervised
     certificate renewal is the goal. Refs #222.
     - Test case and work around for domain names > 64 octets. Fixes #227.
     When the first DNS name of an MD is longer than 63 octets, the certificate
     request will not contain a CN field, but leave it up to the CA to choose one.
     Currently, Lets Encrypt looks for a shorter name in the SAN list given and
     fails the request if none is found. But it is really up to the CA (and what
     browsers/libs accept here) and may change over the years. That is why
     the decision is best made at the CA.
     - Retry delays now have a random +/-[0-50]% modification applied to let
     retries from several servers spread out more, should they have been
     restarted at the same time of day.
     - Fixed several places where the 'badNonce' return code from an ACME server
     was not handled correctly. The test server 'pebble' simulates this behaviour
     by default and helps nicely in verifying this behaviour. Thanks, pebble!
     - Set the default `MDActivationDelay` to 0. This was confusing to users that
     new certificates were deemed not usably before a day of delay. When clocks are
     correct, using a new certificate right away should not pose a problem.
     - When handling ACME authorization resources, the module no longer requires
     the server to return a "Location" header, as was necessary in ACMEv1.
     Fixes #216.
     - Fixed a theoretical uninitialized read when testing for JSON error responses
     from the ACME CA. Reported at <https://bz.apache.org/bugzilla/show_bug.cgi?id=64297>.
     - ACME problem reports from CAs that include parameters in the Content-Type
     header are handled correctly. (Previously, the problem text would not be
     reported and retries could exist CA limits.)
     - Account Update transactions to V2 CAs now use the correct POST-AS-GET method.
     Previously, an empty JSON object was sent - which apparently LE accepted,
     but others reject.

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1887337 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
log tags, my nemesis

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1887340 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
Use ASN1_STRING_data() if openssl verison < 1.1.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1887342 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
refrain from handling ip address alt names in pre 1.1 openssl

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1887343 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
lets try ASN1_STRING_data() for openssl 1.0.2

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1887360 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Changes from trunk regarding the new `ap_ssl_*` functions and hooks.

icing committed:
Using the new ap_ssl_conn_is_ssl() and ap_ssl_var_lookup() in all internal modules.

 * leaving mod_nw_ssl and mod_ssl itself untouched
 * removing mod_ssl.h includes where no longer necessary
 * some modules might skip post_config hooks, but those were left in, even when empty now.

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1887364 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

Commit:
mod_md: synchronized with git and apache trunk including r1888006.

icing committed:
Fixing mod_lua to use new http_ssl.h header file as well.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1888084 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

Commit:
Integrated commit 587d17015167f442c98efaf497504fe3825a3fe7

*) core: provide ap_ssl_* functions in new http_ssl.h header file.

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

icing committed:
*) core/mod_ssl/mod_md: adding OCSP response provisioning as core feature. This
     allows modules to access and provide OCSP response data without being tied
     of each other. The data is exchanged in standard, portable formats (PEM encoded
     certificates and DER encoded responses), so that the actual SSL/crypto
     implementations used by the modules are independant of each other.
     Registration and retrieval happen in the context of a server (server_rec)
     which modules may use to decide if they are configured for this or not.
     The area of changes:
     1. core: defines 2 functions in include/http_ssl.h, so that modules may
        register a certificate, together with its issuer certificate for OCSP
        response provisioning and ask for current response data (DER bytes) later.
        Also, 2 hooks are defined that allow modules to implement this OCSP
        provisioning.
     2. mod_ssl uses the new functions, in addition to what it did already, to
        register its certificates this way. If no one is interested in providing
        OCSP, it falls back to its own (if configured) stapling implementation.
     3. mod_md registers itself at the core hooks for OCSP provisioning. Depending
        on configuration, it will accept registrations of its own certificates only,
        all certficates or none.

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1888723 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
update the damn log tags.

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1888724 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
mod_md: fixed a conversion warning, bumped version number.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1888726 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

Commit:
Merge r1888729 from trunk:

server/protocol.c: remove old hooks links that had been moved to server/ssl.c

rpluem commented on modules/lua/mod_lua.c:
```diff
@@ -1688,15 +1685,12 @@ static const char *register_lua_root(cmd_parms *cmd, void *_cfg,
-        return (const char *)lua_ssl_val(p, s, c, r, (char *)var);
-    }
-    return NULL;
+    return ap_ssl_var_lookup(p, s, c, r, (char *)var);
```
`ap_ssl_var_lookup` expects `const char *` as last parameter, `var` is `const char *`. What is the purpose of converting to `char *`?

rpluem commented on modules/proxy/mod_proxy.c:
```diff
@@ -2866,23 +2860,14 @@ PROXY_DECLARE(int) ap_proxy_ssl_engine(conn_rec *c,
-    }
-    else
-        return NULL;
+    return ap_ssl_var_lookup(p, s, c, r, (char *)var);
```
`ap_ssl_var_lookup` expects `const char *` as last parameter, `var` is `const char *`. What is the purpose of converting to `char *`?

rpluem commented on server/ssl.c:
```diff
@@ -0,0 +1,192 @@
+                                           const char *name)
+{
+    const char *val = ap_run_ssl_var_lookup(p, s, c, r, name);
+    if (val == NULL && module_ssl_is_https) {
```
I guess should be `module_ssl_var_lookup` instead of `module_ssl_is_https`.

rpluem commented on CHANGES:
```diff
@@ -45,6 +137,37 @@ Changes with Apache 2.4.47
+       'ssl_answer_challenge' where modules like mod_md can provide providers
+       to the above mentioned functions.
+     [Stefan Eissing]
+
```
These changes are added twice.

rpluem commented on include/ap_mmn.h:
```diff
@@ -554,14 +554,20 @@
+ * 20120211.104 (2.4.47-dev) Move ap_ssl_* into new http_ssl.h header file
+ * 20120211.105 (2.4.47-dev) Add `ap_bytes_t` to httpd.h.
+ *                         Add ap_ssl_ocsp* hooks and functions to http_ssl.h.
  */
```
Nitpick: Whitespace formating: Before `Add`one space less, before `and` and the last `Add` one space more.

rpluem commented on modules/http2/h2_h2.c:
```diff
@@ -445,44 +437,26 @@ apr_status_t h2_h2_init(apr_pool_t *pool, server_rec *s)
          */
-        val = opt_ssl_var_lookup(pool, s, c, NULL, (char*)"SSL_PROTOCOL");
+        val = ap_ssl_var_lookup(pool, s, c, NULL, (char*)"SSL_PROTOCOL");
         if (val && *val) {
```
`ap_ssl_var_lookup` expects `const char *` as last parameter, `var` is `const char *`. What is the purpose of converting to `char *`?

rpluem commented on modules/http2/h2_h2.c:
```diff
@@ -501,7 +475,7 @@ int h2_is_acceptable_connection(conn_rec *c, request_rec *r, int require_all)
         /* Check TLS cipher blacklist
          */
-        val = opt_ssl_var_lookup(pool, s, c, NULL, (char*)"SSL_CIPHER");
+        val = ap_ssl_var_lookup(pool, s, c, NULL, (char*)"SSL_CIPHER");
```
`ap_ssl_var_lookup` expects `const char *` as last parameter, `var` is `const char *`. What is the purpose of converting to `char *`?

rpluem commented on modules/ssl/ssl_engine_kernel.c:
```diff
@@ -2316,32 +2316,53 @@ void ssl_callback_Info(const SSL *ssl, int where, int rc)
                       "Failed to configure challenge certificate %s",
                       servername);
-        return APR_EGENERAL;
+        rv = APR_EGENERAL; goto cleanup;
```
Nitpick: Style: Can we have two lines here?

rpluem commented on modules/ssl/ssl_engine_kernel.c:
```diff
@@ -2316,32 +2316,53 @@ void ssl_callback_Info(const SSL *ssl, int where, int rc)
                       ERR_error_string(ERR_peek_last_error(), NULL), 
                       servername);
-        return APR_EGENERAL;
+        rv = APR_EGENERAL; goto cleanup;
```
Nitpick: Style: Can we have two lines here?

rpluem commented on modules/ssl/ssl_engine_kernel.c:
```diff
@@ -2316,32 +2316,53 @@ void ssl_callback_Info(const SSL *ssl, int where, int rc)
                       "Challenge certificate and private key %s "
                       "do not match", servername);
-        return APR_EGENERAL;
+        rv = APR_EGENERAL; goto cleanup;
```
Nitpick: Style: Can we have two lines here?

rpluem commented on modules/ssl/ssl_engine_kernel.c:
```diff
@@ -2376,15 +2394,6 @@ static apr_status_t init_vhost(conn_rec *c, SSL *ssl, const char *servername)
-                    return APR_EGENERAL;
-                }
-                SSL_set_verify(ssl, SSL_VERIFY_NONE, ssl_callback_SSLVerify);
-            }
```
Why is this no longer needed?

icing replied:
This was a challenge situation in the ACMEv1 protocol for the alpn https verification. That used a new hostname which led to people being able to get certificates for domain they do not own in shared hosting setups. It has been disabled ever since that was found out.

Since the new mod_md no longer supports ACMEv1, this seems superfluous to keep in the server.

rpluem replied:
Thanks for the clarification

rpluem commented on modules/ssl/ssl_util_ssl.c:
```diff
@@ -511,3 +511,77 @@ char *modssl_SSL_SESSION_id2sz(IDCONST unsigned char *id, int idlen,
+
+    in = BIO_new_mem_buf(cert_pem, -1);
+    if (in == NULL) {
+        rv = APR_ENOMEM; goto cleanup;
```
Nitpick: Style: Can we have two lines here? Here in the remainder of the function?

rpluem reviewed (commented):
Reviewed core and ssl.

icing commented:
The `(char*)` casts are my mistake. I am fine with changing the nitpicks. Will update this, probably today.

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

Commit:
Merge of 2d2e2aee89760f2da93b1a2d89723d65bd241b9c from trunk (r1889009).

    core/ap_ssl_*: changes after review by rpluem
     - removed no longer needed (char*) casts when looking
       up ssl variables.
     - move 'goto cleanup;' on separate source line
     - fixed check for wrong optional function in ap_run_ssl_var_lookup
     - remove ap_bytes_t again from httpd.h and passes now ocsp
       identifier as separate const char* and apr_size_t. This
       follows more how such data is passed in the rest of the
       server.

rpluem commented on CHANGES:
```diff
@@ -93,6 +93,18 @@ Changes with Apache 2.4.47
+
+  *) MPM event/worker: Fix possible crash in child process on early signal
+     delivery.  PR 64533.  [Ruediger Pluem]
+
```
Best thing is to leave all change entries out that do not belong to this backport. In an ideal world they would all get backported by files in `change-entries` and would be added later (latest with the roll script) by executing `make update-changes`.

icing replied:
Agreed. With the recent back and forth about this, I was not sure what the PR should carry.

rpluem replied:
Can you please remove the `CHANGES` entries and put the changes for this backport in a file below `change-entries`? Once done I would vote +1 such that we could get this in 2.4.48.

icing replied:
Did the changes. Agree with you that this is the preferred way now.

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

icing committed:
mod_md: make sure its post-config hook runs before mod_watchdog. if
  not, mod_watchdog thinks it has nothing to do, if no one else installs
  any watchdog instance.

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1889101 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

Commit:
Separated the additions to CHANGES into 2 files underneath change-entries.

Commit:
remove a merge fragment in ap_mmn.h.

bigio commented on server/ssl.c:
```diff
@@ -0,0 +1,193 @@
+                                  conn_rec *c, request_rec *r,
+                                  const char *name)
+{
+    /* Someone retrieved the optional function., not knowning about the
```
```suggestion
    /* Someone retrieved the optional function, not knowing about the
```

bigio commented on server/ssl.c:
```diff
@@ -0,0 +1,193 @@
+                                  const char *name)
+{
+    /* Someone retrieved the optional function., not knowning about the
+     * new API. We redirect them to what they should have inoked. */
```
```suggestion
     * new API. We redirect them to what they should have invoked. */
```

bigio commented on changes-entries/ocsp_stapling_core.txt:
```diff
@@ -0,0 +1,20 @@
+        OCSP, it falls back to its own (if configured) stapling implementation.
+     3. mod_md registers itself at the core hooks for OCSP provisioning. Depending
+        on configuration, it will accept registrations of its own certificates only,
+        all certficates or none.
```
```suggestion
        all certificates or none.
```

bigio commented on modules/md/md_reg.h:
```diff
@@ -112,15 +113,15 @@ apr_status_t md_reg_update(md_reg_t *reg, apr_pool_t *p,
+                                   const md_t *md, struct md_pkey_spec_t *spec, apr_pool_t *p);
 
 /**
  * Synchronise the give master mds with the store.
```
```suggestion
 * Synchronize the given master mds with the store.
```

bigio commented on modules/ssl/ssl_util_ssl.c:
```diff
@@ -511,3 +511,81 @@ char *modssl_SSL_SESSION_id2sz(IDCONST unsigned char *id, int idlen,
+
+/*  _________________________________________________________________
+**
+**  Certficate/Key Stuff
```
```suggestion
**  Certificate/Key Stuff
```

bigio commented on modules/ssl/ssl_util_ssl.h:
```diff
@@ -82,7 +82,26 @@ char       *modssl_SSL_SESSION_id2sz(IDCONST unsigned char *, int, char *, int);
 char *modssl_bio_free_read(apr_pool_t *p, BIO *bio);
-    
+
+/* Read a single certificate and its private key from the give string in PEM format.
```
```suggestion
/* Read a single certificate and its private key from the given string in PEM format.
```

bigio commented on server/ssl.c:
```diff
@@ -0,0 +1,193 @@
+
+static int ssl_is_https(conn_rec *c)
+{
+    /* Someone retrieved the optional function., not knowning about the
```
```suggestion
    /* Someone retrieved the optional function, not knowing about the
```

bigio commented on server/ssl.c:
```diff
@@ -0,0 +1,193 @@
+static int ssl_is_https(conn_rec *c)
+{
+    /* Someone retrieved the optional function., not knowning about the
+     * new API. We redirect them to what they should have inoked. */
```
```suggestion
     * new API. We redirect them to what they should have invoked. */
```

icing committed:
* Spelling fixes after review by @bigio.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1889788 13f79535-47bb-0310-9956-ffa450edef68

Commit:
Merge branch '2.4.x' into icing/2.4.x-ap_ssl_things

icing commented:
Has been merged into subversion 2.4.x branch by r1889793.
